internal/provider/resource_vault_secrets_app.go

```diff
@@ -0,0 +1,169 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	sharedmodels "github.com/hashicorp/hcp-sdk-go/clients/cloud-shared/v1/models"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	clients "github.com/hashicorp/terraform-provider-hcp/internal/clients"
+)
+
+func NewVaultSecretsAppResource() *resourceVaultsecretsApp {
+	return &resourceVaultsecretsApp{}
+}
+
+type resourceVaultsecretsApp struct {
+	client *clients.Client
+}
+
+func (r *resourceVaultsecretsApp) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_vault_secrets_app"
+}
+
+func (r *resourceVaultsecretsApp) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			// TODO: Add validators
+			"app_name": schema.StringAttribute{
+				Required: true,
+			},
+			"description": schema.StringAttribute{
+				Optional: true,
+			},
+		},
+	}
+}
+
+func (r *resourceVaultsecretsApp) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	client, ok := req.ProviderData.(*clients.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	r.client = client
+}
+
+type App struct {
+	ID          types.String `tfsdk:"id"`
+	AppName     string       `tfsdk:"app_name"`
+	Description string       `tfsdk:"description"`
```

Comment:
This is a [common gotcha](https://github.com/hashicorp/terraform-provider-cloudinit/pull/103) for the [conversion rules](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/types/object#setting-values) of `tfsdk` tags for Go types. When you run code like:
```go
var plan App
diags := req.Plan.Get(ctx, &plan)
```

If `AppName` or `Description` is unknown, you'll get an error like:
```bash
An unexpected error was encountered trying to build a value...
Received unknown value, however the target type cannot handle unknown values. Use the corresponding types package type or a custom type that handles unknown values.
```

General rule of thumb, if it has a `tfsdk` tag, it's best to use the Framework [`types`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.4.0/types) package so you can handle unknown and null values without unexpected errors. (like you've done with `ID` 👍🏻 )

Reply:
ahh! this is good to know. Thanks!